sympy/matrices/matrices.py

```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        combined, p = self.LUdecomposition_Simple(iszerofunc=_iszero)
-        L = self.zeros(self.rows)
-        U = self.zeros(self.rows)
-        for i in range(self.rows):
-            for j in range(self.rows):
-                if i > j:
-                    L[i, j] = combined[i, j]
-                else:
-                    if i == j:
-                        L[i, i] = 1
-                    U[i, j] = combined[i, j]
+
+        combined, p = self.LUdecomposition_Simple(iszerofunc=iszerofunc,
+                                                  rankcheck=rankcheck)
+
+        L = self.zeros(combined.rows)
+        U = self.zeros(combined.rows, combined.cols)
+
+        # L is lower triangular self.rows x self.rows
+        # U is upper triangular self.rows x self.cols
+        # L has unit diagonal. For each column in combined, the subcolumn
+        # below the diagonal of combined is shared by L.
+        # If L has more columns than combined, then the remaining subcolumns
+        # below the diagonal of L are zero.
+        # The upper triangular portion of L and combined are equal.
+
+        max_index = min(self.rows, self.cols)
+
+        # Set diagonal entry to 1, and subdiagonal entries of L to
+        # subdiagonal entries of combined
+        for col_i in range(max_index):
+            L[col_i, col_i] = 1
+            for row_i in range(col_i + 1, L.rows):
+                L[row_i, col_i] = combined[row_i, col_i]
+                row_i += 1
+
+        # Fill in the remaining diagonal entries of L
+        for col_i in range(max_index, L.cols):
+            L[col_i, col_i] = 1
+
+        # Set upper triangular portion of U to upper triangular portion of
+        # combined
+        for row_i in range(max_index):
+            for col_i in range(row_i, U.cols):
+                U[row_i, col_i] = combined[row_i, col_i]
```

Comment:
No plans AFAIK.  I could see how that would be more efficient for sparse matrices, but how would it help for dense matrices?  Are you imagining something like

```
def row_gen(k):
    return [f(i) for i in range(k)] + [0]*(n-k)
```

?  How much faster is this than calling a function on every entry?  I'm not sure how you can generate entries and avoid a function call to every entry in general.  Nomatter what happens `simpify` is still called on every entry, so there is always `n*m` overhead with creating a matrix.